src/registry/mod.rs

```
@@ -109,17 +109,18 @@ impl Registry {
 
     /// Fetch the manifest of the OCI object referenced by the given url.
     /// The url is expected to be in the "registry://" format.
-    pub(crate) async fn manifest(
+    pub async fn manifest(
         &self,
         url: &str,
-        sources: &Option<Sources>,
+        sources: Option<&Sources>,
     ) -> Result<oci_distribution::manifest::OciManifest> {
         let url = Url::parse(url).map_err(|_| anyhow!("invalid URL: {}", url))?;
         let reference =
             Reference::from_str(url.as_ref().strip_prefix("registry://").unwrap_or_default())?;
 
         let registry_auth = Registry::auth(reference.registry(), self.docker_config.as_ref());
-        let cp = crate::client_protocol(&url, &sources.clone().unwrap_or_default())?;
+        let sources: Sources = sources.cloned().unwrap_or_default();
```

Given `manifest` returns a `Result`, I'd bubble the error up (or map the error) if `cloned()` fails instead of silently defaulting the structure.

---

NVM, I just realized `cloned()` just returns an `Option` on `Option`. I see, is a way to not consume the `Option` because `unwrap_or_default()` consumes it, right?

In that case, discard the rest of my comments in the same direction.

---

yes, that's the case. I'll resolve the other comments you left

---

doh, indeed, now I see it too. Thanks both!